### ISSUE TITLE
[ Improvement ] Use GPU for model calibration

### ISSUE THREAD

**Issue by NicolasGensollen** [aramis-lab/leaspy#48]
In GitLab by @little_nemo on Oct 5, 2021, 16:26

Preliminary experiments suggest that model calibration can benefit from using GPU backends. This appears to be true especially on “large” datasets (a notion that is still to be defined). Here is a table summarizing the first experiments conducted. All experiments ran on a Xeon E5-2699 @ 2.30GHz machine, equipped with a Quadro M4000 GPU, and consisted in fitting a logistic model on diverse datasets. 
I have not systematically studied the variability of those results between multiple runs for the longer ones, but I am confident that the improvements are significative on the bigger databases. Variability is given as (mean ± std).

| Specification | True ADNI data | Simulated ADNI-like data (1 source) | Simulated ADNI-like data (3 sources) | CEGEDIM real data (small) | CEGEDIM real data (big) |
| ----- | :---: | :---: | :----: | :---: | :----: |
| Subject number | 524 | 5000 | 5000 | 250 | 1000 |
| Average number of visits | 6.25 | 6.25 | 10 | 180 | 190 |
| Feature number | 4 | 4 | 4 | 10 | 10 |
| Source dimension | 3 | 1 | 3 | 1 | 3 |
| Loss used during calibration | MSE | MSE | MSE | crossentropy | crossentropy |
| CPU calibration (1000 iterations) | 2min 14s ± 1.81 (7 runs) | 2min ± 0.5s (7 runs)| 2min 52s ± 3.35 s (7 runs) | 9min 10s | 2h55min |
| GPU calibration (1000 iterations) | **40.9 s** ± 1.57 s (7 runs)| **35.7 s** ± 1.09 s | **46.3 s** ± 1.55 s (7 runs) | **3min 37s** ± 440 ms | **20min** |

I suggest we make it easier to embed computations on GPU, for instance by adding an `on_gpu`-like flag in the parameters available to the user.

The first step is conceptually simple and mostly involves specifying the right device to store the right tensors when the flag is set. 

To this end, one question still needs to be addressed: should “being on the GPU” be an attribute of the `Model` or of the `Algorithm`? To me, the second option is the best one. Indeed, storing `Model` attributes on the GPU would have us refactor a lot of the external pieces of code that we already use. For instance, most plotting utilities tend to expect data stored on the CPU, so checks and “device-passing” should be added everywhere to keep those tools transparent to use. Conversely using an `Algorithm.on_gpu` flag would make it transparent by default, since data would be moved to the GPU only during the calibration step, and moved back to the CPU afterwards.

Further work could / should include:
 * Check that the results are consistent between CPU and GPU calibrations (I don't this should be an issue, but we still need to make sure that this is the case)
 * Benchmarking the algorithms to identify the comfort zones of both CPU and GPU implementation (as in: explicit when either the CPU or GPU backend is more suited, depending on the dataset profile for instance)
 * Identify potential bottlenecks and rewrite the associated computations to take advantage of GPU architecture (low priority since it appears that naively using the GPU backend is already worth the switch for some datasets)

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 12, 2021, 17:12

Could you add to the issue the numerous evidence you and others grasped from embedding Leaspy computations on GPU backends? So we could have everything at the same place.

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Oct 12, 2021, 17:15

I do agree with you to have (if necessary) a GPU flag related to the `Algorithm` settings, not to the model.

Prior (or in parallel of) GPU usage, the one that will handle this issue should first check whether all the "tensorization" that could be made in the model are done. It basically means that the code should rely on matrix computations as frequently as possible.

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Oct 13, 2021, 13:46

Yes, I just did that thank you! I also added a few more experiments to demonstrate how the calibration behaves on different tasks. Do not hesitate to ask if you would like to see specific settings / datasets / models to be tested. Hopefully I will clean my “proof-of-concept” branch soon enough so that anyone will be able to run such benchmarks, but in the meantime I can run a few more experiments depending on what you would like to see.

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Oct 19, 2021, 18:24

One issue that needs to be addressed / discussed in the general context of using GPUs is the testing pipelines for the GPU computations. I have not written such tests for the moment, but I can think of at least two points that needs to be considered:

* I am not sure that GitLab gives “free access” to GPUs to run testing pipelines on its servers; I will check but it would be surprising to me. This would be a bit troublesome for the CI.
* Tests involving GPUs cannot be run locally on most laptops and part of the workstations, that do not have a CUDA-compatible GPU. Therefore they should be distinct from the tests triggered by `python -m unittest discover tests/{unit, func}_tests`.

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Mar 1, 2022, 10:15

MR https://gitlab.com/icm-institute/aramislab/leaspy/-/merge_requests/56 having been merged, I think I can close this issue for the moment. There might still be potential improvements to study (related to vectorization, etc), but could be discussed again in another issue when someone has some time to dedicate to this.
